src/controllers/post.js

```diff
@@ -0,0 +1,115 @@
+const Post = require('../models/post');
+
+exports.getAllPosts = async (_, res, next) => {
+  try {
+    const posts = await Post.find();
+    res.json(posts);
+  } catch (err) {
+    next(err);
+  }
+};
+
+exports.getOnePost = async (req, res, next) => {
+  const { id } = req.params;
+  try {
+    const post = await Post.findById(id);
+    if (!post) {
+      res
+        .status(422)
+        .json({ message: "the post you are looking for wasn't found" });
+    } else {
+      res.json(post);
+    }
+  } catch (err) {
+    next(err);
+  }
+};
+
+exports.addPost = async (req, res, next) => {
+  const postData = req.body;
+  try {
+    const newPost = await Post.create(postData);
+    res.status(201).json(newPost);
+  } catch (err) {
+    next(err);
+  }
+};
+
+exports.filterPosts = async (req, res, next) => {
+  const { sender, title } = req.query;
+  if (!sender && !title) {
+    res
+      .status(400)
+      .json({ message: 'make sure you send a valid query parameter' });
+  } else {
+    const query = {};
+    if (sender) query.sender = sender;
+    if (title) query.title = title;
+    try {
+      const posts = await Post.find(query);
+      if (posts.length === 0) {
+        res.status(422).json({ message: 'No posts by these parameters' });
```

Comment:
I think we can remove the status code 422 here, since no posts found is not necessarily an error, just a result based on the search params.